src/treeview/writer.js

```diff
@@ -0,0 +1,623 @@
+			const nodeAfterPriority = this._priorities.get( nodeAfter );
+
+			// Do not merge same nodes with different priorities.
+			if ( nodeBeforePriority === undefined || nodeBeforePriority !== nodeAfterPriority ) {
+				return Position.createFromPosition( position );
+			}
+
+			// Move all children nodes from node placed after selection and remove that node.
+			const count = nodeBefore.getChildCount();
+			nodeBefore.appendChildren( nodeAfter.getChildren() );
+			nodeAfter.remove();
+
+			// New position is located inside the first node, before new nodes.
+			// Call this method recursively to merge again if needed.
+			return this.mergeAttributes( new Position( nodeBefore, count ) );
+		}
+
+		return position;
+	}
+
+	/**
+	 * Insert node or nodes at specified position. Takes care about breaking attributes before insertion
+	 * and merging them afterwards.
+	 *
+	 * @param {core.treeView.Position} position Insertion position.
+	 * @param {core.treeView.Node|Iterable.<core.treeView.Node>} nodes Node or nodes to insert.
+	 * @returns {core.treeView.Range} Range around inserted nodes.
+	 */
+	insert( position, nodes ) {
+		const container = this.getParentContainer( position );
+		const insertionPosition = this.breakAttributes( position );
+
+		const length = container.insertChildren( insertionPosition.offset, nodes );
+		const endPosition = insertionPosition.getShiftedBy( length );
+		const start = this.mergeAttributes( insertionPosition );
+
+		// When no nodes were inserted - return collapsed range.
+		if ( length === 0 ) {
+			return new Range( start, start );
+		} else {
+			// If start position was merged - move end position.
+			if ( !start.isEqual( insertionPosition ) ) {
+				endPosition.offset--;
+			}
+			const end = this.mergeAttributes( endPosition );
+
+			return new Range( start, end );
+		}
+	}
+
+	/**
+	 * Removes provided range from the container.
+	 *
+	 * Throws {@link core.CKEditorError CKEditorError} `treeview-writer-invalid-range-container` when
+	 * {@link core.treeView.Range#start start} and {@link core.treeView.Range#end end} positions are not placed inside
+	 * same parent container.
+	 *
+	 * @param {core.treeView.Range} range Range to remove from container. After removing, it will be updated
+	 * to a collapsed range showing the new position.
+	 * @returns {Array.<core.treeView.Node>} The array of removed nodes.
+	 */
+	remove( range ) {
+		// Range should be placed inside one container.
+		if ( this.getParentContainer( range.start ) !== this.getParentContainer( range.end ) ) {
+			/**
+			 * Range is not placed inside same container.
+			 *
+			 * @error treeview-writer-invalid-range-container
+			 */
+			throw new CKEditorError( 'treeview-writer-invalid-range-container' );
+		}
+
+		// If range is collapsed - nothing to remove.
+		if ( range.isCollapsed ) {
+			return [];
+		}
+
+		// Break attributes at range start and end.
+		const breakRange = this.breakRange( range );
```

Comment:
Why not destructuring? :D
`const { start: breakStart, end: breakEnd } = this.breakRange( range );`